rsocket-core/src/main/java/io/rsocket/framing/SetupFrame.java

```diff
@@ -0,0 +1,460 @@
+    }
+
+    byteBuf = byteBuf.writeByte(getLengthAsUnsignedByte(metadataMimeType));
+    byteBuf = Unpooled.wrappedBuffer(byteBuf, metadataMimeType.retain(), byteBufAllocator.buffer());
+
+    byteBuf = byteBuf.writeByte(getLengthAsUnsignedByte(dataMimeType));
+    byteBuf = Unpooled.wrappedBuffer(byteBuf, dataMimeType.retain(), byteBufAllocator.buffer());
+
+    byteBuf = appendMetadata(byteBufAllocator, byteBuf, metadata);
+    byteBuf = appendData(byteBuf, data);
+
+    return RECYCLER.get().setByteBuf(byteBuf);
+  }
+
+  /**
+   * Returns the data MIME-type encoding as a UTF-8 {@link String}.
+   *
+   * @return the data MIME-type encoding as a UTF-8 {@link String}
+   */
+  public String getDataMimeTypeAsUtf8() {
+    return mapDataMimeType(byteBuf -> byteBuf.toString(UTF_8));
+  }
+
+  /**
+   * Returns the keep alive interval.
+   *
+   * @return the keep alive interval
+   */
+  public Duration getKeepAliveInterval() {
+    return Duration.ofMillis(getByteBuf().getInt(OFFSET_KEEPALIVE_INTERVAL));
+  }
+
+  /**
+   * Returns the major version of the protocol.
+   *
+   * @return the major version of the protocol
+   */
+  public int getMajorVersion() {
+    return getByteBuf().getUnsignedShort(OFFSET_MAJOR_VERSION);
+  }
+
+  /**
+   * Returns the max lifetime.
+   *
+   * @return the max lifetime
+   */
+  public Duration getMaxLifetime() {
+    return Duration.ofMillis(getByteBuf().getInt(OFFSET_MAX_LIFETIME));
+  }
+
+  /**
+   * Returns the metadata MIME-type encoding as a UTF-8 {@link String}.
+   *
+   * @return the metadata MIME-type encoding as a UTF-8 {@link String}
+   */
+  public String getMetadataMimeTypeAsUtf8() {
+    return mapMetadataMimeType(byteBuf -> byteBuf.toString(UTF_8));
+  }
+
+  /**
+   * Returns the minor version of the protocol.
+   *
+   * @return the minor version of the protocol
+   */
+  public int getMinorVersion() {
+    return getByteBuf().getUnsignedShort(OFFSET_MINOR_VERSION);
+  }
+
+  /**
+   * Returns the resume identification token as a UTF-8 {@link String}.
+   *
+   * @return the resume identification token as a UTF-8 {@link String}
+   */
+  public Optional<String> getResumeIdentificationTokenAsUtf8() {
+    return mapResumeIdentificationToken(byteBuf -> byteBuf.toString(UTF_8));
+  }
+
+  /**
+   * Returns whether the lease flag is set.
+   *
+   * @return whether the lease flag is set
+   */
+  public boolean isLeaseFlagSet() {
+    return isFlagSet(FLAG_LEASE);
+  }
+
+  @Override
+  public <T> T mapData(Function<ByteBuf, T> function) {
+    return mapData(getMetadataLengthOffset(), function);
+  }
+
+  /**
+   * Exposes the data MIME-type for mapping to a different type.
+   *
+   * @param function the function to transform the data MIME-type to a different type
+   * @param <T> the different type
+   * @return the data mapped to a different type
+   * @throws NullPointerException if {@code function} is {@code null}
+   */
+  public <T> T mapDataMimeType(Function<ByteBuf, T> function) {
```

Comment:
Can you explain the thinking over map functions for something as trivial as mime type.  it's generally a US-ASCII string.

Reply:
Generally is the problem:

> MIME Type for encoding of Data and Metadata. This SHOULD be a US-ASCII string that includes...

The use of “SHOULD” leaves the possibility of a non-US-ASCII string for this value.  I do believe that everyone will use the [`String` variants](https://github.com/nebhale/rsocket-java/blob/31ce73bef70cf8c5d83a0990250d55366d2c7807/rsocket-core/src/main/java/io/rsocket/framing/SetupFrame.java#L284) but wanted to be pedantically accurate to the spec.

Reply:
I think maybe making the default easy to find default return a UTF-8 string, and possibly having this mapDataMimeType as a backup.  But TBH I think the SHOULD argument was over US-ASCII vs UTF-8 previously, not over arbitrary binary encodings here.

Reply:
What would you do the method name `getDataMimeTypeAsUtf8()`?  I'm not sure what would make one method easier to find than the other.

Reply:
Or rather, is the bug here that it should be `toString(US_ASCII)`?